test/app-luatest/key_estimator_test.lua

```diff
@@ -0,0 +1,187 @@
+    t.assert_equals(est:estimate(), 1)
+    est:add({'bla'})
+    t.assert_equals(est:estimate(), 1)
+    est:add({'bla', 'bla', 'bla..'})
+    t.assert_equals(est:estimate(), 1)
+    est:add({'another bla'})
+    t.assert_equals(est:estimate(), 2)
+
+    t.assert_error_msg_equals("Invalid tuple format.",
+        key_estimator.add, est, {0})
+end
+
+g.test_hll_dense_estimation_error = function()
+    local fmt = key_def.new({{type = "string", fieldno = 1}})
+
+    local min_prec = key_estimator.MIN_PRECISION
+    local max_prec = key_estimator.MAX_PRECISION
+    for prec = min_prec, max_prec, 1 do
+        local est = key_estimator.new(fmt, prec)
+        local std_err = hll_std_error(prec)
+
+        --- cardinality for wich the HyperLogLog algorithm will be used.
+        local card = 5 * hll_n_counters(prec)
+        for i = 1, card, 1 do
+            est:add({tostring(i)})
+        end
+
+        --- 99.7% to lie in expected interval. See the comment to hll_std_error.
+        local margin = card * std_err * 3
+        t.assert_almost_equals(est:estimate(), card, margin)
+    end
+end
+
+g.test_hll_sparse_estimation_error = function()
+    --- Since the precision parameter defines only the maximal size of the
+    --- sparse representation there is no need in tests for all available
+    --- precision values.
+    local max_prec = key_estimator.MAX_PRECISION
+    local fmt = key_def.new({{type = "number", fieldno = 1}})
+    local est = key_estimator.new(fmt, max_prec, key_estimator.SPARSE)
+    --- Error for the sparse representation can't be computed but this error
+    --- must be less than error for any available precision.
+    local std_err = hll_std_error(max_prec + 1)
+
+    --- Sparse representation stores 4-bite pairs instead of 6-bit counters,
+    --- so 6/32<6 times fewer pairs can be stored in the same amount of memory.
+    local max_card = hll_n_counters(key_estimator.MAX_PRECISION) / 6
+    local n_steps = key_estimator.MAX_PRECISION -
+                    key_estimator.MIN_PRECISION + 1
+    local card_step = max_card / n_steps
+
+    for n = 1, n_steps, 1 do
+        for i = 1, card_step, 1 do
+            est:add({n*card_step + i})
+        end
+        local card = n * card_step
+        --- 99.7% to lie in expected interval. See the comment to hll_std_error.
+        local margin = card * std_err * 3
+        t.assert_almost_equals(est:estimate(), card, margin)
+    end
+end
+
+g.test_merge = function()
+    local fmt = key_def.new({{type = "string", fieldno = 1}})
+
+    local min_prec = key_estimator.MIN_PRECISION
+    local max_prec = key_estimator.MAX_PRECISION
+    for prec = min_prec, max_prec, 1 do
+        local est1 = key_estimator.new(fmt, prec, key_estimator.DENSE)
+        local est2 = key_estimator.new(fmt, prec, key_estimator.DENSE)
+        local std_err = hll_std_error(prec)
+
+        --- cardinality for wich the HyperLogLog algorithm will be used.
+        local card = 5 * hll_n_counters(prec)
```

Comment:
No, by explaining the formula I mean, where does `5` come from?

Reply:
Added a more detailed description.